sdk/src/azure/iot/az_iot_adu_client.c

```diff
@@ -0,0 +1,956 @@
+  _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+      ref_json_writer, AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_INTERFACE_ID)));
+  _az_RETURN_IF_FAILED(az_json_writer_append_string(
+      ref_json_writer, AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_INTERFACE_ID)));
+
+  _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+      ref_json_writer, AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_ADU_VERSION)));
+  _az_RETURN_IF_FAILED(
+      az_json_writer_append_string(ref_json_writer, device_properties->adu_version));
+
+  if (!az_span_is_content_equal(
+          device_properties->delivery_optimization_agent_version, AZ_SPAN_EMPTY))
+  {
+    _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+        ref_json_writer, AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_DO_VERSION)));
+    _az_RETURN_IF_FAILED(az_json_writer_append_string(
+        ref_json_writer, device_properties->delivery_optimization_agent_version));
+  }
+
+  _az_RETURN_IF_FAILED(az_json_writer_append_end_object(ref_json_writer));
+
+  /* Fill the compatibility property names. */
+  _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+      ref_json_writer,
+      AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_COMPAT_PROPERTY_NAMES)));
+  _az_RETURN_IF_FAILED(az_json_writer_append_string(
+      ref_json_writer, client->_internal.options.device_compatibility_properties));
+
+  /* Add last installed update information */
+  if (last_install_result != NULL)
+  {
+    _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+        ref_json_writer,
+        AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_LAST_INSTALL_RESULT)));
+    _az_RETURN_IF_FAILED(az_json_writer_append_begin_object(ref_json_writer));
+
+    _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+        ref_json_writer, AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_RESULT_CODE)));
+    _az_RETURN_IF_FAILED(
+        az_json_writer_append_int32(ref_json_writer, last_install_result->result_code));
+
+    _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+        ref_json_writer,
+        AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_EXTENDED_RESULT_CODE)));
+    _az_RETURN_IF_FAILED(
+        az_json_writer_append_int32(ref_json_writer, last_install_result->extended_result_code));
+
+    if (!az_span_is_content_equal(last_install_result->result_details, AZ_SPAN_EMPTY))
+    {
+      // TODO: Add quotes if result_details is not enclosed by quotes.
+      _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+          ref_json_writer, AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_RESULT_DETAILS)));
+      _az_RETURN_IF_FAILED(
+          az_json_writer_append_string(ref_json_writer, last_install_result->result_details));
+    }
+
+    for (int32_t i = 0; i < last_install_result->step_results_count; i++)
+    {
+      az_span remaining_buffer = get_json_writer_remaining_buffer(ref_json_writer);
```

Comment:
Not for now, but we should get the Azure SDK team opinion about this... Seems a little unsettled to just grab space from the end of the buffer.

Reply:
I'll add a TODO on it